service-worker.js

```diff
@@ -0,0 +1,188 @@
+      return caches.delete(cacheName);
+    }
+  });
+  await Promise.all(deletePromises);
+}
+
+self.addEventListener("activate", async (event) => {
+  console.log("Activating service worker.");
+  await clearOldCaches();
+  clients.claim();
+});
+
+const ASSETS_REQUEST_URL_REGEX = /^\/(.*-)?(?<docId>\w+)\/assets\/(?<path>.*)$/;
+
+self.addEventListener("fetch", async (event) => {
+  const url = new URL(event.request.url);
+
+  const match = url.pathname.match(ASSETS_REQUEST_URL_REGEX);
+  if (match) {
+    const { docId, path } = match.groups;
+
+    const automergeUrl = `automerge:${docId}`;
+    if (!isValidAutomergeUrl(automergeUrl)) {
+      event.respondWith(
+        new Response(`Invalid document id ${docId}`, {
+          status: 404,
+          headers: { "Content-Type": "text/plain" },
+        })
+      );
+      return;
+    }
+
+    event.respondWith(
+      (async () => {
+        const handle = (await repo).find(automergeUrl);
+        await handle.whenReady();
+        const doc = await handle.doc();
+
+        if (!doc) {
+          return new Response(
+            `Document unavailable.\n${automergeUrl}: ${handle.state}`,
+            {
+              status: 500,
+              headers: { "Content-Type": "text/plain" },
+            }
+          );
+        }
+
+        if (!doc.assetsDocUrl) {
+          return new Response(`Document has no assets doc.\n${automergeUrl}`, {
+            status: 404,
+            headers: { "Content-Type": "text/plain" },
+          });
+        }
+
+        const assetsHandle = (await repo).find(doc.assetsDocUrl);
+        await assetsHandle.whenReady();
+        const assetsDoc = await assetsHandle.doc();
+
+        if (!assetsDoc || !assetsDoc.files) {
+          return new Response(
+            `Document unavailable.\n${assetsHandle.url}: ${assetsHandle.state}`,
+            {
+              status: 500,
+              headers: { "Content-Type": "text/plain" },
+            }
+          );
+        }
+
+        const decodedPath = decodeURI(path);
+        const file = assetsDoc.files[decodedPath];
+
+        if (!file) {
+          return new Response(
+            `Not found\npath: ${JSON.stringify(
+              decodedPath
+            )}\nfiles:${JSON.stringify(Object.keys(assetsDoc.files), null, 2)}`,
+            {
+              status: 404,
+              headers: { "Content-Type": "text/plain" },
+            }
+          );
+        }
+
+        if (file.contentType) {
+          return new Response(file.contents, {
+            headers: { "Content-Type": file.contentType },
+          });
+        }
+
+        return new Response(
+          `Invalid file entry.\n${
+            assetsHandle.url
+          }:\nfileEntry:${JSON.stringify(file)}`,
+          {
+            status: 404,
+            headers: { "Content-Type": "text/plain" },
+          }
+        );
+      })()
+    );
+  }
+  // disable caching other files for now
```

Comment:
should we just remove this code for now?

Reply:
Maybe we could add a condition to only enable it in production. If it leads to issues, we could still remove it. Otherwise we still don't have real offline support